[FLINK-14495][docs] Add documentation for memory control of RocksDB state backend (#10498)
apache/flink: docs/ops/state/large_state_tuning.md

of the TaskManagers by the same amount. Not doing that may result in YARN/Mesos/etc terminating the JVM processes for
allocating more memory than configured.

#### Bound total memory usage of RocksDB instance(s) per slot

RocksDB allocates native memory without control of JVM, and might lead the process to exceed total memory budget of the container to get killed in container environment (e.g. Kubernetes).
From Flink-1.10, we provide a solution to limit total memory usage for RocksDb instance(s) per slot by leveraging RocksDB's mechanism to 
share [cache](https://github.com/facebook/rocksdb/wiki/Block-Cache) and [write buffer manager](https://github.com/facebook/rocksdb/wiki/Write-Buffer-Manager) among instance(s).
Generally speaking, we mainly have three parts of memory usage for RocksDB in Flink scenario: block cache, index & bloom filters and memtables 
(refer to [memory-usage-in-rocksdb](https://github.com/facebook/rocksdb/wiki/Memory-usage-in-RocksDB)).
The basic idea is to share a `Cache` object with desired capacity among all RocksDB instances, 

> and [cost memory used in memtable to that cache](https://github.com/facebook/rocksdb/wiki/Write-Buffer-Manager#cost-memory-used-in-memtable-to-block-cache) via write buffer manager.

Reviewer: As much as I find this interesting I think this is to in the weeds for most users. This kind of implementation detail is better for javadocs and maybe also a blog post.

Besides, we also cache index & filters into that cache, then the major use of memory would be well capped.

> There exist two ways to enable this feature:

Reviewer: What do you think about something like this for the first paragraph? 

> RocksDB allocates native memory outside of the JVM, which could lead the process to exceed the total memory budget. This can be especially problematic in containerized environments such as Kubernetes that kill processes that exceed their memory budgets. Flink can limit total memory usage for each RocksDB instance per slot by leveraging shareable [caches](https://github.com/facebook/rocksdb/wiki/Block-Cache) and [write buffer manager's](https://github.com/facebook/rocksdb/wiki/Write-Buffer-Manager) among all instances in a single slot. These shared caches will place a hard upper limit on the [three components](https://github.com/facebook/rocksdb/wiki/Memory-usage-in-RocksDB) that use the majority of memory when RocksDB is deployed as a state backend: block cache, index and bloom filters, and MemTables.

Author: Thanks for your suggestion, I will take this description with correctting something not so accurate.

  -  Turn `state.backend.rocksdb.memory.managed` as true. If so, RocksDB state backend will use the managed memory budget of the task slot to set the capacity of that shared cache object.
  -  Configure the memory size of `state.backend.rocksdb.memory.fixed-per-slot` to set the fixed total amount of memory per slot. 

>   This option will override `state.backend.rocksdb.memory.managed` option when configured.

Reviewer: Can you expand on when a user should use one configuration over another? Could we maybe phrase this as a basic vs advanced configuration?

> We also provide two parameters to tune the memory fraction of memtable and index & filters:

Reviewer: We -> Flink
Also the end of this sentence "memtable and index and filters" sounds like it's missing a word but I'm not sure what. 
```suggestion
Flink also provides two parameters to tune the memory fraction of MemTable and index & filters:
```

  - `state.backend.rocksdb.memory.write-buffer-ratio`, by default `0.5`. If RocksDB memory bounded feature is turned on, 50% of memory size would be used by write buffer manager by default.
  - `state.backend.rocksdb.memory.high-prio-pool-ratio`, by default `0.1`.
  If RocksDB memory bounded feature is turned on, 10% 0f memory size would be set as high priority for index and filters in shared block cache by default. 
  By enabling this, index and filters would not need to compete against data blocks for staying in cache to minimize performance problem if those index and filters are evicted by data blocks frequently.

>   Moreover, we also pin L0 level filter and index into cache by default to mitigate performance problem, more details could refer to [RocksDB-doc](https://github.com/facebook/rocksdb/wiki/Block-Cache#caching-index-filter-and-compression-dictionary-blocks).

Reviewer: Another way to remove "we" that it doesn't get repetitive. 
```suggestion
  Moreover, the L0 level filter and index are pinned into the cache by default to mitigate performance problems, more details could refer to the [RocksDB documentation](https://github.com/facebook/rocksdb/wiki/Block-Cache#caching-index-filter-and-compression-dictionary-blocks).
```

*Experts only*: To control memory manually instead of using managed memory, user can set `state.backend.rocksdb.memory.managed` as `false` and control via `ColumnFamilyOptions`.
Or to save some manual calculation, through the `state.backend.rocksdb.memory.fixed-per-slot` option which will override `state.backend.rocksdb.memory.managed` when configured.
With the later method, please tune down `taskmanager.memory.managed.size` or `taskmanager.memory.managed.fraction` to `0` 
and increase `taskmanager.memory.task.off-heap.size` by "`taskmanager.numberOfTaskSlots` * `state.backend.rocksdb.memory.fixed-per-slot`" accordingly.

#### Tune performance when bounding RocksDB memory usage.

There might existed performance regression compared with previous no-memory-limit case if you have too many states per slot.
- If you observed this behavior and not running jobs in containerized environment or don't care about the over-limit memory usage.
The easiest way to wipe out the performance regression is to disable memory bound for RocksDB, e.g. turn `state.backend.rocksdb.memory.managed` as `false`.

> Moreover, please refer to [memory configuration migration guide](WIP) to know how to keep backward compatibility to previous memory configuration.

Author: @azagrebin @xintongsong Please fill in the link to memory configuration migration guide after that guide have been merged.

Reviewer 2: I don't think we need to add this information now since the migration guide is still WIP. Let's open new PR for necessary update after the in-progress work is done instead of blocking this PR.

Author: Yes, this PR is not blocked by their PR but tell them here leaves a blank to be filled when they're merging their PR. This was what I want to tell and already edited my previous comment.

Reviewer: I agree with Yu. Please remove the line and comment on the migration guide ticket asking them to add the link on their PR.

Author: @carp84 @sjwiesman , okay I have removed that line and will remind them at their PR.

Reviewer 3: The memory config PR is open (#10999). You can either rebase on it or comment if this PR gets merged first.

Reviewer 3: The docs were reworked in the subsequent PR #10987, where do we want to refer the main memory config page now from?

Reviewer 2: After the rework the reference to migration guide seems not that necessary to me, but if we still would like to mention it, I think the most suitable place is line#132 in `large_state_tuning.md` (or simply search for `state.backend.rocksdb.memory.managed: false` in this file)

allocating more memory than configured.

### Bounding RocksDB Memory Usage

RocksDB allocates native memory outside of the JVM, which could lead the process to exceed the total memory budget.
This can be especially problematic in containerized environments such as Kubernetes that kill processes who exceed their memory budgets.
Flink limit total memory usage of RocksDB instance(s) per slot by leveraging shareable [cache](https://github.com/facebook/rocksdb/wiki/Block-Cache)
and [write buffer manager](https://github.com/facebook/rocksdb/wiki/Write-Buffer-Manager) among all instances in a single slot by default.
The shared cache will place an upper limit on the [three components](https://github.com/facebook/rocksdb/wiki/Memory-usage-in-RocksDB) that use the majority of memory
when RocksDB is deployed as a state backend: block cache, index and bloom filters, and MemTables.
This feature is enabled by default and could be controlled by two ways:
  -  Integrate with managed memory of task manager: turn `state.backend.rocksdb.memory.managed` as true. If so, RocksDB state backend will use the managed memory budget of the task slot to set the capacity of that shared cache object.

>   This operation is enabled by default, which means Flink would always choose to integrate RocksDB memory usage with the managed memory first.

Author: I think [comment](https://github.com/apache/flink/pull/10498#discussion_r355570820) from Seth might be reasonable, we should avoid specific Flink version in documentation.

>   This operation is enabled by default, which means Flink would always choose to integrate RocksDB memory usage with the managed memory first.

Reviewer: Yes, the documentation is versioned so we don’t need to specify versions inline. 

Just say:  

This feature is enabled by default along with managed memory. Flink will use the managed memory budget as the per-slot memory limit for RocksDB state backend(s).

>   This operation is enabled by default, which means Flink would always choose to integrate RocksDB memory usage with the managed memory first.

Reviewer 2: Fair enough to remove the version information. However, I think it's still necessary to simplify the description here.

  -  Not integrated with managed memory: configure the memory size of `state.backend.rocksdb.memory.fixed-per-slot` to set the fixed total amount of memory per slot.
  This option will override `state.backend.rocksdb.memory.managed` option when configured and ignore calculated managed memory per slot from task manager.

>   User could also configure `taskmanager.memory.task.off-heap.size` to set additional quota in off-heap memory, which should be equal to `taskmanager.numberOfTaskSlots` * ``state.backend.rocksdb.memory.fixed-per-slot``, to fit in Flink's memory model.

Author: I think user could manually control memory usage via `ClumnFamilyOptions` even our feature of shared cache is not introduced in. What's more, after our feature is introduced, if `state.backend.rocksdb.memory.managed` is still `true` and the `ClumnFamilyOptions` which added by user would not take real effect since we will finally create a write buffer manager and a cache to share.
I think we might not need to talk about the `ClumnFamilyOptions`
I

>   User could also configure `taskmanager.memory.task.off-heap.size` to set additional quota in off-heap memory, which should be equal to `taskmanager.numberOfTaskSlots` * ``state.backend.rocksdb.memory.fixed-per-slot``, to fit in Flink's memory model.

Reviewer 2: I think the context here is already not using the managed memory or `state.backend.rocksdb.memory.managed` is `false`. On the other hand, to make it explicit, we could also add a simple line like "*Experts Only*: To control memory manually instead of using managed memory, you can set `state.backend.rocksdb.memory.managed` to `false` and control through `ClumnFamilyOptions`. Or to save some manual calculation, .... "

  If RocksDB memory bounded feature is turned on, 10% 0f memory size would be set as high priority for index and filters in shared block cache by default.
  By enabling this, index and filters would not need to compete against data blocks for staying in cache to minimize performance problem if those index and filters are evicted by data blocks frequently.
  Moreover, the L0 level filter and index are pinned into the cache by default to mitigate performance problems,
  more details could refer to the [RocksDB-documentation](https://github.com/facebook/rocksdb/wiki/Block-Cache#caching-index-filter-and-compression-dictionary-blocks).

<span class="label label-info">Note</span> The shared `cahe` and `write buffer manager` will override customized settings of block cache and write buffer via `PredefinedOptions` and `OptionsFactory`.

#### Tune performance when bounding RocksDB memory usage.

There might existed performance regression compared with previous no-memory-limit case if you have too many states per slot.
If you observed this behavior and not running jobs in containerized environment or does not care about the over-limit memory usage.

> The easiest way to wipe out the performance regression is to disable memory bound for RocksDB, e.g. turn `state.backend.rocksdb.memory.managed` as false.

Author: For RocksDB, turn `state.backend.rocksdb.memory.managed` as false is enough. The managed memory size or fraction of task manager would not affect RocksDB anymore.

> The easiest way to wipe out the performance regression is to disable memory bound for RocksDB, e.g. turn `state.backend.rocksdb.memory.managed` as false.

Reviewer 2: True but then the managed memory won't be used by anyone and just left in waste, right?

> The easiest way to wipe out the performance regression is to disable memory bound for RocksDB, e.g. turn `state.backend.rocksdb.memory.managed` as false.

Author: If user does not use RocksDB, the managed memory is still wasted. I am not sure that part of memory is in the scope of this PR.

> The easiest way to wipe out the performance regression is to disable memory bound for RocksDB, e.g. turn `state.backend.rocksdb.memory.managed` as false.

Reviewer 2: ok, I agree this is out of the scope of this PR, let's discuss this in other PRs when necessary.

Otherwise you need to increase the upper memory for RocksDB:
  - Increase the managed memory size of `taskmanager.memory.managed.size` or the fraction via `taskmanager.memory.managed.fraction`.
  - Increase the memory size of `state.backend.rocksdb.memory.fixed-per-slot`, which is not integrated with managed memory of task manager, as well as increasing `taskmanager.memory.task.off-heap.size` accordingly.

Apart from increasing total memory, you could also tune some options:
  - Decrease the arena block size, which is 1/8 of write-buffer size by default. This targets to decrease the possibility of turning mutable mem-table as immutable when write buffer manager reserve memory at the granularity of arena block.
  Increase the max background flush threads for each DB instance. This targets to flush immutable mem-tables as fast as possible to not block or stall writes.

>   User could use options factory below to tune:

Reviewer 2: I'm not sure whether we need to emphasize these two options (especially for the "bounding RocksDB memory usage" case), for two reasons:
1. For non-expert users, it's hard for them to understand and may cause confusion. Too many suggestions means no suggestion for non-experts, IMHO.
2. For expert users who know well about RocksDB, these are not something new. They may already tune these (and even more) options in their environment.

Author: I prefer to leave them here since this documentation is already be named as "tuning".

Reviewer 2: The information given here are all RocksDB internals. As a user, I could hardly tell in what case I should decrease the arena block size or increase the background flush threads.

What's more, why do these two configurations value most? Why not tune other configurations? Are we giving the suggestion based on our own experimental case? Does it fit for all cases?

I insist that giving recommendation on tuning Flink settings instead of RocksDB ones would help more.
